File: day1/solution.js
// improve the time-space complexity



function SortSquarredArray(array){
    if (array.length > 0) {
        const newArr = new Array(array.length).fill(0);
        let pointerLeft = 0;
        let pointerRight = array.length - 1;

        for (let i = array.length - 1; i >= 0; i--) {
            const leftSquared = array[pointerLeft] * array[pointerLeft];
            const rightSquared = array[pointerRight] * array[pointerRight];

            if (leftSquared > rightSquared) {
                newArr[i] = leftSquared;
                pointerLeft++;
            } else {
                newArr[i] = rightSquared;
                pointerRight--;
            }
        }
        return newArr;
    } else {
        return [];
    }
}



const array_testcase2 = [-3 , -2, 1, 5,4] 
const b = [1,4,5,7]
const c = [-6,-1,0,5,10]
console.log(SortSquarredArray(b))
console.log(SortSquarredArray(array_testcase2))
console.log(SortSquarredArray([]))

File: day1/solution.py
def sort_squared_array(array):
    if len(array) > 0:
        new_arr = [0] * len(array)
        pointer_left = 0
        pointer_right = len(array) - 1

        for i in range(len(array) - 1, -1, -1):
            left_squared = array[pointer_left] ** 2
            right_squared = array[pointer_right] ** 2

            if left_squared > right_squared:
                new_arr[i] = left_squared
                pointer_left += 1
            else:
                new_arr[i] = right_squared
                pointer_right -= 1

        return new_arr
    else:
        return []

# Test cases
array_testcase2 = [-3, -2, 1, 5, 4]
b = [1, 4, 5, 7]
c = [-6, -1, 0, 5, 10]

print(sort_squared_array(b))  # [1, 16, 25, 49]
print(sort_squared_array(array_testcase2))  # [1, 4, 9, 16, 25]
print(sort_squared_array([]))  # []
